File: apps/android/app/src/main/java/org/llamenos/hotline/ui/settings/ErasureViewModel.kt
package org.llamenos.hotline.ui.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.llamenos.hotline.api.ApiService
import javax.inject.Inject

data class ErasureRequest(
    val id: String = "",
    val status: String = "",
    val requestedAt: String? = null,
    val executeAt: String? = null,
)

data class ErasureUiState(
    val activeRequest: ErasureRequest? = null,
    val isLoading: Boolean = false,
    val isMutating: Boolean = false,
    val error: String? = null,
    val success: String? = null,
    val showRequestConfirmation: Boolean = false,
    val showCancelConfirmation: Boolean = false,
)

/**
 * ViewModel for user-facing erasure request lifecycle.
 *
 * Provides load/request/cancel operations for the self-service erasure
 * flow accessible from Settings. Uses StateFlow for Compose state collection.
 */
@HiltViewModel
class ErasureViewModel @Inject constructor(
    private val apiService: ApiService,
) : ViewModel() {

    private val _uiState = MutableStateFlow(ErasureUiState())
    val uiState: StateFlow<ErasureUiState> = _uiState.asStateFlow()

    fun loadStatus() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val response = apiService.request<Map<String, Any?>>("GET", "/api/erasure/me")
                _uiState.update {
                    it.copy(
                        isLoading = false,
                        activeRequest = parseErasureResponse(response),
                    )
                }
            } catch (e: Exception) {
                // 404 means no active request — that's fine
                if (e.message?.contains("404") != true) {
                    _uiState.update { it.copy(isLoading = false, error = e.message) }
                } else {
                    _uiState.update { it.copy(isLoading = false, activeRequest = null) }
                }
            }
        }
    }

    fun showRequestConfirmation() {
        _uiState.update { it.copy(showRequestConfirmation = true) }
    }

    fun dismissRequestConfirmation() {
        _uiState.update { it.copy(showRequestConfirmation = false) }
    }

    fun requestErasure() {
        viewModelScope.launch {
            _uiState.update { it.copy(isMutating = true, error = null, success = null) }
            try {
                val response = apiService.request<Map<String, Any?>>("POST", "/api/erasure/me")
                _uiState.update {
                    it.copy(
                        isMutating = false,
                        activeRequest = parseErasureResponse(response),
                        showRequestConfirmation = false,
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isMutating = false, error = e.message) }
            }
        }
    }

    fun showCancelConfirmation() {
        _uiState.update { it.copy(showCancelConfirmation = true) }
    }

    fun dismissCancelConfirmation() {
        _uiState.update { it.copy(showCancelConfirmation = false) }
    }

    fun cancelErasure() {
        viewModelScope.launch {
            _uiState.update { it.copy(isMutating = true, error = null, success = null) }
            try {
                apiService.requestNoContent("DELETE", "/api/erasure/me")
                _uiState.update {
                    it.copy(
                        isMutating = false,
                        activeRequest = null,
                        showCancelConfirmation = false,
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isMutating = false, error = e.message) }
            }
        }
    }

    private fun parseErasureResponse(response: Map<String, Any?>): ErasureRequest? {
        val request = response["request"] as? Map<String, Any?> ?: return null
        return ErasureRequest(
            id = request["id"] as? String ?: "",
            status = request["status"] as? String ?: "",
            requestedAt = request["requestedAt"] as? String,
            executeAt = request["executeAt"] as? String,
        )
    }
}

File: apps/ios/Sources/ViewModels/ErasureViewModel.swift
import Foundation
import UIKit

/// View model for user-facing erasure request lifecycle.
@Observable
final class ErasureViewModel {
    private let apiService: APIService

    // MARK: - State

    /// Current erasure request status, nil if no active request.
    var activeRequest: ErasureRequest?

    /// Whether the request is loading.
    var isLoading: Bool = false

    /// Whether a mutation (create/cancel) is in progress.
    var isMutating: Bool = false

    /// Error from last operation.
    var errorMessage: String?

    /// Success message.
    var successMessage: String?

    /// Show the confirmation dialog for requesting erasure.
    var showRequestConfirmation: Bool = false

    /// Show the confirmation dialog for cancelling erasure.
    var showCancelConfirmation: Bool = false

    // MARK: - Init

    init(apiService: APIService) {
        self.apiService = apiService
    }

    // MARK: - API

    /// Check current erasure request status.
    func loadStatus() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil

        do {
            let response: ErasureStatusResponse = try await apiService.request(
                method: "GET",
                path: "/api/erasure/me"
            )
            activeRequest = response.request
        } catch {
            // 404 means no active request — that's fine
            if !error.localizedDescription.contains("404") {
                errorMessage = error.localizedDescription
            }
            activeRequest = nil
        }

        isLoading = false
    }

    /// Request account erasure.
    func requestErasure() async {
        isMutating = true
        errorMessage = nil
        successMessage = nil

        do {
            let response: ErasureStatusResponse = try await apiService.request(
                method: "POST",
                path: "/api/erasure/me"
            )
            activeRequest = response.request

            let generator = UINotificationFeedbackGenerator()
            generator.notificationOccurred(.warning)

            successMessage = NSLocalizedString(
                "erasure_status_scheduled",
                comment: "Erasure scheduled"
            )
        } catch {
            errorMessage = error.localizedDescription
        }

        isMutating = false
    }

    /// Cancel a pending erasure request.
    func cancelErasure() async {
        isMutating = true
        errorMessage = nil
        successMessage = nil

        do {
            try await apiService.request(
                method: "DELETE",
                path: "/api/erasure/me"
            )
            activeRequest = nil

            let generator = UINotificationFeedbackGenerator()
            generator.notificationOccurred(.success)

            successMessage = NSLocalizedString(
                "erasure_status_cancelled",
                comment: "Erasure cancelled"
            )
        } catch {
            errorMessage = error.localizedDescription
        }

        isMutating = false
    }

    // MARK: - Countdown

    /// Time remaining until erasure executes (nil if no active request).
    var timeRemaining: TimeInterval? {
        guard let request = activeRequest,
              let executeAt = request.executeAt else { return nil }
        let remaining = executeAt.timeIntervalSinceNow
        return remaining > 0 ? remaining : 0
    }
}

// MARK: - Models

struct ErasureRequest: Codable, Sendable {
    let id: String
    let status: String
    let requestedAt: Date?
    let executeAt: Date?

    var statusDisplay: String {
        switch status {
        case "pending": return NSLocalizedString("erasure_status_pending", comment: "Pending")
        case "scheduled": return NSLocalizedString("erasure_status_scheduled", comment: "Scheduled")
        case "executing": return NSLocalizedString("erasure_status_executing", comment: "Executing")
        case "completed": return NSLocalizedString("erasure_status_completed", comment: "Completed")
        case "cancelled": return NSLocalizedString("erasure_status_cancelled", comment: "Cancelled")
        default: return status
        }
    }
}

struct ErasureStatusResponse: Codable, Sendable {
    let request: ErasureRequest?
}
